### ISSUE TITLE
Support thread: I made a PR before the reformat and it has merge conflicts now! How do I fix it?

### ISSUE THREAD

**Issue by snappizz** [supercollider/supercollider#4428]
This is an official support thread for anyone who has a branch that was created before the big C++ code reformat (#4426, #4425) and is now dealing with merge conflicts.

First off, don't do anything brash. **You don't need to close your PR.** This following process will update a PR that you've already filed.

If you have a branch that contains work prior to the major C++ reformatting commit, just follow these steps. This is a "support thread" so that you can leave comments to ask for help.

## Install requirements

You will need clang v8.0.0 (exactly) and Python 3.6+ (Python 2.7 is also OK but may encounter bugs with non-ASCII characters). Instructions on grabbing the dependencies are here: https://github.com/supercollider/supercollider/wiki/Cpp-formatting-instructions#requirements

## Update your repo

First, update the `3.10` and `develop` branches.

```
git checkout 3.10
git pull upstream 3.10
git checkout develop
git pull upstream develop
```

The next steps also rely on the existence of the tags `tag-clang-format-develop` and `tag-clang-format-3.10`. To grab these, run

```
git pull upstream --tags
```

Remember to switch back to the original branch you were working on:

```
git checkout <branch-name>
```

## 3.10 or develop?

Figure out whether your PR was originally made from 3.10 or develop. If you forgot, look up the PR on GitHub and see what it was filed against. In this case, it's 3.10:

![Untitled](https://user-images.githubusercontent.com/1211064/58756835-d2f8a100-84b4-11e9-8979-e1812a91abb6.png)

## Rebase onto the commit before the reformat

Remember to switch back to the original branch you were working on:

```
git checkout <branch-name>
```

If your branch was based on `develop`, run:

```
git rebase tag-clang-format-develop^  # DON'T FORGET THE ^ !!!!
```

For 3.10:

```
git rebase tag-clang-format-3.10^  # DON'T FORGET THE ^ !!!!
```

The ^, which you absolutely must not forget, indicates that you're rebasing onto the commit immediately before the relevant reformat commit. This gives you access to the `tools/clang-format.py` script.

## Reformat it!

If your branch was based on `develop`, run:

```
tools/clang-format.py rebase -b develop
```

For 3.10:

```
tools/clang-format.py rebase -b 3.10
```

This will switch you over to a new branch called `<branch-name>-reformatted`.

## Confirm

Inspect the changes -- do they all look good?

If so:

```
# rewrite the original branch to match the reformatted one
git branch -f <branch-name>

# Force push your changes to the branch in your fork of SC
git push -f origin <branch-name>
```

The PR will automatically update. **There is no need to close and re-open your PR.**

## Troubleshooting

We are using fairly tricky git features here, and I'm aware a lot of people are relatively new to git. **If something gets screwed up, don't panic** and **don't do reckless things like deleting your entire repository.** It will probably not fix your issue.

Instead, just ask for help! The fastest way to get help is [Slack](https://join.slack.com/t/scsynth/shared_invite/enQtMzk3OTY3MzE0MTAyLWY1ZGE1MTJjYmI5NTRkZjFmNjZmNmYxOWI0NDZkNjdkMzdkNjgxNTJhZGVlOTEwYjdjMDY5OWM0ZTA4NWFiOGY) (since Brian and I are on there), but not everyone is on that, so you can also use this ticket as an official support thread. If you're having a problem, just leave a comment and someone more experienced with the process can help you out.

**Comment by lnihlen** [comment]
one thing on OS X - the easiest way to get ```clang-format``` and ```clang-format-diff.py``` pinned on version 8.0.0. was to download them from the [LLVM Site](https://releases.llvm.org/download.html), unpack and install them somewhere (I put it in ```~/usr/local/```). Then you will have to specify the path to these manually when using the ```tools/clang-format.py``` script, so my command line starts looking like:

```
$ tools/clang-format.py -c ~/usr/local/clang+llvm-8.0.0-x86_64-apple-darwin/bin/clang-format -d ~/usr/local/clang+llvm-8.0.0-x86_64-apple-darwin/share/clang/clang-format-diff.py format
```

It might be useful to provide an option to specify SC-specific environment variables for these paths, something like ```SC_CLANG_FORMAT``` and ```SC_CLANG_FORMAT_DIFF``` might make life easier. Also I noticed a couple of minor doc bugs in the help script for clang-format.py, so maybe I'll take a crack at a PR for these if there's interest.

**Comment by snappizz** [comment]
found a bug in `tools/clang-format.py` :( turns out it throws an error if it encounters a commit in the branch that makes *only* whitespace changes. this can be worked around by doing an interactive rebase and squashing the whitespace-only changes into other commits.

**Comment by brianlheim** [comment]
Two things--

1. I would recommend just adding both the /bin and /share/clang directories from the llvm distro to PATH. This is what Homebrew does when it installs clang, and it seems much easier than doing this via env vars.

2. Making any changes to `clang-format.py rebase` at this point will be dicey unless you really want people to check out the script by itself and put it somewhere else. Following the instructions as they are now, people will always get the version right before the reformat commit. So either we should focus on updating the instructions, OR tell people to get the script separately and focus on updating the script. Since it seems like there are issues with the script that go beyond documentation, I'd vote the latter.

@nathan do you have an example PR for the whitespace only commit issue? I can look into it.

**Comment by lnihlen** [comment]
Ah, so brew does allow for you to install clang with it but by default (and I suspect this will be the case for most folks as it is the process described in the [installation instructions](https://docs.brew.sh/Installation) for brew), homebrew relies on the version of clang installed along with the Xcode command-line tools, which right now is it version 10.0.1, and doesn't include clang-format:

```
MacBook-Pro:supercollider luken$ which clang
/usr/bin/clang
MacBook-Pro:supercollider luken$ which clang-format
MacBook-Pro:supercollider luken$ clang --version
Apple LLVM version 10.0.1 (clang-1001.0.46.4)
Target: x86_64-apple-darwin18.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
MacBook-Pro:supercollider luken$ 
```

While it is possible to get brew to install its own version of clang, and it is (fortunately, for now) at 8.0.0, there's some scary language in there:

```
MacBook-Pro:supercollider luken$ brew info llvm
llvm: stable 8.0.0 (bottled), HEAD [keg-only]
Next-gen compiler infrastructure
https://llvm.org/
Not installed
From: https://github.com/Homebrew/homebrew-core/blob/master/Formula/llvm.rb
==> Dependencies
Build: cmake ✔
Required: libffi ✔, swig ✘
==> Requirements
Build: xcode ✔
==> Options
--HEAD
	Install HEAD version
==> Caveats
To use the bundled libc++ please add the following LDFLAGS:
  LDFLAGS="-L/usr/local/opt/llvm/lib -Wl,-rpath,/usr/local/opt/llvm/lib"

llvm is keg-only, which means it was not symlinked into /usr/local,
because macOS already provides this software and installing another version in
parallel can cause all kinds of trouble.

==> Analytics
install: 17,798 (30 days), 49,350 (90 days), 175,996 (365 days)
install_on_request: 13,605 (30 days), 37,899 (90 days), 135,520 (365 days)
build_error: 0 (30 days)
```

That, coupled with the notion that brew can update their version of clang during the course of normal brew update/upgrade and then break my supercollider workflow, which assumes 8.0.0 for now, makes me want to have a standalone toolchain that I control specifically for clang-format. Hence the local install and the environment variables.

**Comment by brianlheim** [comment]
Hm, ok! That makes sense. And there's another benefit to env vars, which is that you can then also run the CMake targets (which can't take command-line arguments) and `./lint.py`/`./format.py` scripts (which don't take these arguments yet). I'll make the PR for this!

**Comment by lnihlen** [comment]
Hit another snag when trying to rebase the Parser branch:

```
MacBook-Pro:supercollider luken$ git checkout --track origin/topic/paser-lexer-info-in-sclang
Branch 'topic/paser-lexer-info-in-sclang' set up to track remote branch 'topic/paser-lexer-info-in-sclang' from 'origin'.
Switched to a new branch 'topic/paser-lexer-info-in-sclang'
MacBook-Pro:supercollider luken$ git rebase tag-clang-format-develop^
First, rewinding head to replay your work on top of it...
Applying: first draft of parse data emission in supercollider
Applying: checkpoint commit before I start on AST processing
Applying: rework parser output format to account for multiple parents in parse tree
Applying: some tinkering with parsing, broken
Applying: parser now restoring tree shape and simplyfing tree
Applying: start work on building new tree with whitespace
Applying: some progress on new approach and testing
Applying: starting in on documentation
Applying: hit an interesting snag in number parsing
MacBook-Pro:supercollider luken$ tools/clang-format.py -c ~/usr/local/clang+llvm-8.0.0-x86_64-apple-darwin/bin/clang-format -d ~/usr/local/clang+llvm-8.0.0-x86_64-apple-darwin/share/clang/clang-format-diff.py rebase -b develop
--- Formatting e3b432b6fd28fdd273352316680ee3d6ce5b2bbd (1 of 9)
	Skipping file 'HelpSource/Classes/Interpreter.schelp' (no formatting to apply)
	Skipping file 'SCClassLibrary/Common/Core/Kernel.sc' (no formatting to apply)
	Skipping file 'lang/CMakeLists.txt' (no formatting to apply)
  File "tools/clang-format.py", line 601, in main
    rebase_branch(options.clang_format, options.commit1, options.commit2, options.target)
  File "tools/clang-format.py", line 433, in rebase_branch
    new_file.write(file_contents)

*** ERROR:
'ascii' codec can't encode character u'\u2026' in position 629: ordinal not in range(128)

MacBook-Pro:supercollider luken$ git status
HEAD detached at tag-clang-format-develop
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   HelpSource/Classes/Interpreter.schelp

no changes added to commit (use "git add" and/or "git commit -a")
MacBook-Pro:supercollider luken$ python --version
Python 2.7.10
```

Brian is this some of the python versioning stuff you were looking at?

**Comment by lnihlen** [comment]
Answer from Brian on this one is that this is a problem for PRs that touched PyrLexer, because it has non-ASCII characters in it. The solution is to use python3 for the script.

**Comment by brianlheim** [comment]
Sorry again about that. My python skills are not up to the task of figuring this stuff out.

**Comment by dyfer** [comment]
First - thanks to Brain and all the devs for the work on reformat!

I think it would be useful to document the optimal way of installing clang-format on macOS. I personally ended up downloading it from llvm website and manually copying to ```usr/local/bin/``` but that's definitely not a clean solution. If we go for adding ```bin``` from clang distro to the ```$PATH```, where should it be put in the first place?

I also went to update my PR according to the instructions and it went fairly smoothly (thanks again!), with the exception of encountering the same error as @lnihlen . This leaves the user in a bit of a messy state, on detached HEAD. Is addressing that something we should add to the documentation?

In case anyone searches for this, for me the solution was: 
1. discard changes in the current state of the repo
2. checkout the PR branch again
3. run the script with python3: ```$ python3 tools/clang-format.py rebase -b develop```

**Comment by mtmccrea** [comment]
> First - thanks to @brianlheim and all the devs for the work on reformat!

📣Echo!📣 Great work on this, and thank you for keeping us up-to-date along the way and for documenting the process!

I decided to try the easy way—after seeing that the macOS precompiled binary was pretty large—and used homebrew:
```
brew install clang-format
```
because there's only one formula, it went smoothly, and **no need to mess with proper install location**:
```
-> clang-format --version
clang-format version 8.0.0 (tags/google/stable/2019-01-18)
```
In the future if other versions are available, there would need to be a tap for v8.0.0, but for now we're good :)

I proceeded as instructed, updating my `develop` and `3.10` branches, as well as the tags, then:
```
-> git rebase tag-clang-format-develop^
```
So far so good.
```
python3 tools/clang-format.py rebase -b develop
```
Snag:
```
*** ERROR:
Your working tree has pending changes. You must have a clean working tree before proceeding.
```
Turns out that my `nova-simd` submodule was `-dirty`. After some messing around, the fix was simple:
```
-> git submodule foreach --recursive git checkout .
```
Ran `clang-format` again and it worked smoothly! 
I finished up with the remaining instructions—renamed my branch, forced pushed it to `origin`, and the PR updated.

Let the PR party resume! 🥳

**Comment by brianlheim** [comment]
Thanks for the comments and suggestions everyone! I will try to address all of this as soon as possible. Really glad that it's working well so far! Also, credit is very much due to the devs at MongoDB who wrote the original version of the script-- the nastiest logic around rebasing was thankfully figured out by them.

**Comment by brianlheim** [comment]
> I think it would be useful to document the optimal way of installing clang-format on macOS. I personally ended up downloading it from llvm website and manually copying to usr/local/bin/ but that's definitely not a clean solution. If we go for adding bin from clang distro to the $PATH, where should it be put in the first place? I see @lnihlen recommends ~/usr/local/

@dyfer I didn't document this because there are multiple ways to do it and I didn't want to take a side or cause trouble for anyone whose development environment is different from how I manage mine. As pointed out above you can install it (for now) from homebrew. You can also download it from LLVM's site and install it wherever you like; adding it to PATH or not is the user's choice, as you may have other clang versions locally and not want this version to take precedence. I typically install to `/usr/local/llvm` and add `/usr/local/llvm/bin` to PATH. I am still of the opinion that it's best to let developers manage their own installations here. But we could certainly suggest a 'simple solution'. What do you think we should recommend?

> I also went to update my PR according to the instructions and it went fairly smoothly (thanks again!), with the exception of encountering the same error as @lnihlen . This leaves the user in a bit of a messy state, on detached HEAD. Is addressing that something we should add to the documentation?

Yes definitely, I will add it to the wiki page!

**Comment by brianlheim** [comment]
@mtmccrea Glad the script worked for you!

Here's the updated section in the Wiki: https://github.com/supercollider/supercollider/wiki/Cpp-formatting-instructions#specific-issues

**Comment by mtmccrea** [comment]
Update re: Homebrew and Xcode `format` and `lint` tagets.

Despite having successfully rebased and formatted previous pre-formatting PRs:
```
-> python3 tools/clang-format.py rebase -b develop
``` 
I wasn't able to run the formatting scripts on new work:
```
-> python3 tools/clang-format.py format
```
It looks like the Hombrew install of `clang-format` doesn't contain `clang-format-diff.py`. So in the end I had to uninstall it and install the gigantic LLVM package. I placed the whole package in the location suggested by @brianlheim—`/usr/local/llvm` (I created the `llvm` directory)—and added the couple of paths by his suggestion too:

> I would recommend just adding both the `/bin` and `/share/clang` directories from the llvm distro to PATH. 

Then I could format new additions with `clang-format.py`.

**Question**: should the rebasing of old PRs have worked without `clang-format-diff.py`??

Now a new issue... despite being able to run `clang-format.py` from the command line, both the `format` and `lint` Xcode **targets** don't work:
```
PhaseScriptExecution CMake\ Rules build/tools/SuperCollider.build/Debug/format.build/Script-3C8BD2005A3A436D83BCD23B.sh
    cd /Users/admin/src/supercollider-mtmccrea
    /bin/sh -c /Users/admin/src/supercollider-mtmccrea/build/tools/SuperCollider.build/Debug/format.build/Script-3C8BD2005A3A436D83BCD23B.sh

echo ""

/Users/admin/src/supercollider-mtmccrea/tools/clang-format.py format
  File "/Users/admin/src/supercollider-mtmccrea/tools/clang-format.py", line 588, in main
    % CLANG_FORMAT_ACCEPTED_VERSION)

*** ERROR:
Could not find clang-format-diff.py. Please ensure that clang 8.0.0 is installed and that clang-format-diff.py is in your PATH.

make: *** [/Users/admin/src/supercollider-mtmccrea/build/tools/CMakeFiles/format] Error 6
Command /bin/sh failed with exit code 2
```

After digging around in cmake files I can't quite pin down why these targets wouldn't work but running the command directly would... ideas?

**Comment by brianlheim** [comment]
> It looks like the Hombrew install of clang-format doesn't contain clang-format-diff.py.

I see it:

```
[brianheim@BrianMBP supercollider]$ brew install clang-format
[brianheim@BrianMBP supercollider]$ find /usr/local/Cellar/clang-format -name 'clang-format-diff.py'
/usr/local/Cellar/clang-format/2019-01-18/share/clang/clang-format-diff.py
```

Maybe you didn't add it to PATH?

> Question: should the rebasing of old PRs have worked without clang-format-diff.py??

Yes: https://github.com/supercollider/supercollider/wiki/Cpp-formatting-instructions#requirements

> if you want to use the lint or format CMake target, or lint.py or format.py script, you will need clang-format-diff.py

> After digging around in cmake files I can't quite pin down why these targets wouldn't work but running the command directly would... ideas?

It looks like you need to make sure clang-format-diff.py is in PATH when Xcode launches a build target. How are you doing that currently in your terminal? For me, this is the value of PATH when Xcode launched a build target (even if I launched Xcode from my terminal): `'/Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin:/Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/local/bin:/Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/libexec:/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/usr/bin:/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/usr/local/bin:/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/usr/bin:/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/usr/local/bin:/Applications/Xcode-beta.app/Contents/Developer/usr/bin:/Applications/Xcode-beta.app/Contents/Developer/usr/local/bin:/Applications/Xcode-beta.app/Contents/Developer/Tools:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin'`. However, in my terminal, PATH includes many more directories.

I actually symlink clang-format-diff.py into `/usr/local/bin`, so I don't have a problem with this. But if you're just trying to solve this problem, I would recommend setting the `SC_CLANG_FORMAT` and `SC_CLANG_FORMAT_DIFF` env vars in the "Edit Scheme..." dialog, in the Arguments tab (assuming the versions of Xcode we're using are reasonably similar in UI).

**Comment by brianlheim** [comment]
Btw, will try to clarify these things on the wiki page later tonight!

**Comment by mtmccrea** [comment]
Many thanks @brianlheim for sticking with me on this.

> BrianMBP supercollider]$ `find /usr/local/Cellar/clang-format -name 'clang-format-diff.py'`
/usr/local/Cellar/clang-format/2019-01-18/share/clang/clang-format-diff.py

`find` is handy, thanks for the tip. I also found this useful for getting the search location for `find`:
```
-> brew --prefix clang-format
/usr/local/opt/clang-format
```

**Aside:** To explain my confusion a bit: I was running `clang-format.py` when rebasing onto the formatting branch, it ran without errors. So I assumed homebrew had installed everything necessary and in the proper location. When it then broke on newer branches, because it couldn't find `clang-format-diff.py`, this led me to believe that, oh homebrew actually didn't install it, and the rebasing onto the older formatting branch worked because it didn't require `clang-format-diff.py`. In hindsight it could be that there just weren't any changes that needed formatting in those branches? Hence my question: should the rebasing of old PRs have worked **without** clang-format-diff.py?? Anyway...

>Maybe you didn't add it to PATH?

I had forgotten that `-> export PATH=$PATH:` doesn't persist across sessions, so that was a dumb flub. Once I added `/usr/local/Cellar/clang-format/2019-01-18/share/clang/` [to PATH permanently](https://www.architectryan.com/2012/10/02/add-to-the-path-on-mac-os-x-mountain-lion/#.Uydjga1dXDg), then I was good to go. (Not sure if it's preferred to add via `~/.bash_profile` instead?).

>if you're just trying to solve this problem, I would recommend setting the SC_CLANG_FORMAT and SC_CLANG_FORMAT_DIFF env vars in the "Edit Scheme..." dialog, in the Arguments tab

I didn't have any luck with this, I tried setting the vars to the scripts's paths as well as to their parent folders and couldn't get either to work so I just did...

>symlink `clang-format-diff.py` into `/usr/local/bin`

and that did it for me 👍

>It looks like you need to make sure `clang-format-diff.py` is in `PATH` when Xcode launches a build target. 

This was key. Be sure to close out Xcode, do all the necessary linking, rebuild the Xcode project (`cmake -G Xcode -DCMAKE_PREFIX_PATH=`brew --prefix qt5`  ..`), then it should work after that.

With any luck, I've covered all possible dumb mistakes necessary for this thread 🙄
